lib/cucumber/formatter/json.rb

```diff
@@ -1,19 +1,280 @@
+          data = encode64(content)
+        else
+          if mime_type =~ /;base64$/
+            mime_type = mime_type[0..-8]
+            data = src
+          else
+            data = encode64(src)
+          end
+        end
+        test_step_embeddings << { mime_type: mime_type, data: data }
+      end
+
+      private
+
+      def same_feature_as_previous_test_case?(feature)
+        current_feature[:uri] == feature.file && current_feature[:line] == feature.location.line
+      end
+
+      def first_step_after_background?(test_step)
+        test_step.source[1].name != @element_hash[:name]
+      end
+
+      def prepare_world_hook?(test_step)
+        test_step.source.last.location.file.end_with?('cucumber/filters/prepare_world.rb')
+      end
+
+      def hook?(test_step)
+        hook_source?(test_step.source.last)
+      end
+
+      def hook_source?(step_source)
+        ['Before hook', 'After hook', 'AfterStep hook'].include? step_source.name
+      end
+
+      def current_feature
+        @feature_hash ||= {}
+      end
+
+      def feature_elements
+        @feature_hash[:elements] ||= []
+      end
+
+      def steps
+        @element_hash[:steps] ||= []
+      end
+
+      def hooks_of_type(test_step)
+        name = test_step.source.last.name
+        if name == 'Before hook'
+          return before_hooks
+        elsif name == 'After hook'
+          return after_hooks
+        elsif name == 'AfterStep hook'
+          return after_step_hooks
+        else
+          fail 'Unkown hook type ' + name
+        end
+      end
+
+      def before_hooks
+        @element_hash[:before] ||= []
+      end
+
+      def after_hooks
+        @element_hash[:after] ||= []
+      end
+
+      def after_step_hooks
+        @step_hash[:after] ||= []
+      end
+
+      def test_step_output
+        @step_or_hook_hash[:output] ||= []
+      end
+
+      def test_step_embeddings
+        @step_or_hook_hash[:embeddings] ||= []
+      end
+
+      def create_step_hash(step_source)
+        step_hash = {
+          keyword: step_source.keyword,
+          name: step_source.name,
+          line: step_source.location.line
+        }
+        step_hash[:doc_string] = create_doc_string_hash(step_source.multiline_arg) if step_source.multiline_arg.doc_string?
+        step_hash
+      end
+
+      def create_doc_string_hash(doc_string)
+        {
+          value: doc_string.content,
+          content_type: doc_string.content_type,
+          line: doc_string.location.line
+        }
+      end
+
+      def add_match_and_result(step_source, result)
+        @step_or_hook_hash[:match] = create_match_hash(step_source, result)
+        @step_or_hook_hash[:result] = create_result_hash(result)
+      end
+
+      def create_match_hash(step_source, result)
+        if result.undefined? || hook_source?(step_source)
+          location = step_source.location
+        else
+          location = @runtime.step_match(step_source.name).file_colon_line
```

Comment:
For step_definition test_steps I first tried to get the location from the action of the test_step, but for step definition the location is always the line in `Cucumber::StepMatch` where the action is created from the step definitions, so I instead did what `LegacyApi::Adapter` did - asking the `Cucumber::Runtime::SupportCode` to get the StepMatch (with the correct location). When I came to hook, I had to "solve the problem the right way" - but then test_steps from step_definitions and hook are treated differently. With new support in Cucumber-Ruby-Core (https://github.com/cucumber/cucumber-ruby-core/commit/c595f23366ee1c2c4997bc734f8bf7cd53fed151) actions know the location of the step_definition of hook that they are based on, and test_steps can be queried for the location of their action, so no `if` statement is necessary any more (to pass the location of the step_definition when creating the action we easy when the support in Cucumber-Ruby-Core was in place 48c0f34).